### PR TITLE
Added coarse rescaling for waterfall plots

### PR DESCRIPTION
**Pull request by LamarMoore**
Fixes issue #13629 
## Notes

The reason why waterfall plots were not being rescaled was because the offset curves were shifted at the point of being drawn as opposed to modifying the data before plotting (see PlotCurve::DrawCurve). This discrepancy resulted in graph rescaling only taking original data into account as opposed to the new position of the shifted curve (curve no longer fit within scale window). 

I have added slight work around so that the graph limits are modified if a waterfall plot is in use. However, I am not sure this is the best solution.
## Testing

Run the following script and then right click the workspace **Plot Spectrum**, select the **WaterFall Plot** checkbox and click **Plot All**:

``` python
import math
X = []
Y = []
for i in range(0,100):
       x = (i-50)/20.
       X.append(i)
       Y.append(math.exp(-x*x))
ws = CreateWorkspace(X+X,Y+Y,NSpec=2)
```

### CODE REVIEW

**Comment by raquelalvarezbanos** [comment]
Works well with the example script. However, I tried with a real data set and it didn't work:

```
Load(Filename=r'D:\mantid\mantid-build\ExternalData\Testing\Data\UnitTest\MUSR00022725.nxs', OutputWorkspace='MUSR00022725', SpectrumMin=1, SpectrumMax=4)
```
Plot all the spectra, now "Plot type" > "Waterfall", set the offsets to:

Total Y Offset: 60%
Total X Offset: 10%

The rescaling does not work. I think it is because only the last curve is considered when calculating the maximum values.

**Comment by LamarMoore** [comment]
I have added some changes which allow rescaling of the graph to be native to Qwt plot i.e. the offset of the waterfall plot is added to the curve data as opposed to being offset in the qpainter when the curve is drawn. However, this means that PlotCurve (which directly inherits from QwtPlotCurve) needs to be aware of Mantid workspaces and data types. I am not sure this is what we want.

**Comment by raquelalvarezbanos** [comment]
This branch has merge conflicts, could you resolve them?

**Comment by raquelalvarezbanos** [comment]
While testing the last changes, I got the plot below (Y offset 20%):

new

In contrast to (same Y offset, master branch):

old

**Comment by martyngigg** [comment]
As a minor point, can you update the initial description so that it uses one of the recognised keywords for closing issues: https://help.github.com/articles/closing-issues-via-commit-messages/. "Addresses issue" won't do anything at the moment

File: MantidPlot/src/Mantid/MantidUI.cpp [hunk 1]
```diff
@@ -3230,6 +3230,7 @@ MultiLayer* MantidUI::plot1D(const QMultiMap<QString,int>& toPlot, bool spectrum
     g->checkValuesInAxisRange(firstCurve);
   }
   ml->toggleWaterfall(waterfallPlot);
+  g->setAutoScale();
```
Comment (martyngigg) [inline MantidPlot/src/Mantid/MantidUI.cpp]: Is this going to override whether autoscaling is enabled/disabled by the user preferences?

A few lines above (#L3220) there is a comment to that effect.

File: MantidPlot/src/PlotCurve.cpp [hunk 2]
```diff
@@ -278,75 +297,109 @@ void PlotCurve::computeWaterfallOffsets()
...
+    if (index > 0 && c) {
+      // Compute offsets based on the maximum value for the curve
+      double x_offset = index * g->waterfallXOffset() * 0.01 *
+                        g->curve(0)->maxXValue() / (double)(curves - 1);
+      double y_offset = index * g->waterfallYOffset() * 0.01 *
+                        g->curve(0)->maxYValue() / (double)(curves - 1);
+
+      // Clone the workspace so that the data is not corrupted when applying
+      // offsets
+      std::unique_ptr<MatrixWorkspace> wksp_clone =
```
Comment (martyngigg) [inline MantidPlot/src/PlotCurve.cpp]: I don't think this is the best approach to the problem. You are correct that we don't want to affect the original data but if I have a large dataset, plot a few curves then display a waterfall plot I wouldn't expect my memory requirements to double.

**Comment by raquelalvarezbanos** [comment]
I've tested the functionality again and works as expected (in addition, it looks better for the muon dataset I was using as a test :))

**Pull request merged**